Documentation for RTD

jamadden opened the pull request:
Fixes #4

Plus some minor whitespace adjustments.

Docs are viewable at http://zopeviewlet.readthedocs.io/en/rtd-docs/

jamadden committed:
Documentation for RTD

Fixes #4

Plus some minor whitespace adjustments.

mgedmin commented on MANIFEST.in:
```diff
@@ -8,4 +8,7 @@ include .coveragerc
 recursive-include src *
 
+recursive-include docs *rst
+recursive-include docs *py
```
Looks weird without the `.`s, but  I suppose it works.

mgedmin commented on src/zope/viewlet/README.rst:
```diff
@@ -616,8 +630,9 @@ absolute URL for it:
-then one javascript resource file:
+There is also a :obj:`javascript viewlet base class
+<.JavaScriptBundleViewlet>` which knows how to render more then one
+javascript resource file:
```
Typo (in the original): then -> than

jamadden replied:
Sure enough!

mgedmin commented on src/zope/viewlet/manager.py:
```diff
@@ -155,16 +217,23 @@ class WeightOrderedViewletManager(ViewletManagerBase):
+        """
+        Sort the viewlets based on their ``weight`` attribute (if present;
+        viewlets without a ``weight`` are sorted at the beginning but are
+        otherwise  unordered).
```
Double space.

jamadden replied:
Thank you. I'm using [a new comment formatter](https://github.com/glyph/python-docstring-mode) that's RST and doctest aware, and it usually works great, but it does seem to have an issue of inserting extra spaces sometimes.

mgedmin commented on src/zope/viewlet/viewlet.py:
```diff
@@ -96,6 +104,11 @@ class ResourceViewletBase(object):
+        """
+        Lookup the resource from our path using the
+        :class:`++resource++ namespace <zope.traversing.namespace.resource>`
+        and call it, returning the results.
```
"The results" will be the URL of that resource, since that's what you get when you call resources.

jamadden replied:
To nitpick a bit, that may be what you get when you call a `zope.browserresource.interfaces.IResource`, but that's not guaranteed by the (default) ++resource namespace. The (default) resource namespace, as implemented in `zope.traversing.namespace.resource`, simply finds the default named adapter for the request. The interface that adapter implements is not specified. *This* object puts a further (previously implicit) constraint upon it that it is callable. This object doesn't even require that the results of calling it be a string.

This object is further derived with objects like JavaScriptViewlet that use templates and assume that getURL returns something that is usable as a string in a TALES context. Perhaps those objects could use more documentation, and **certainly** the common usage and interaction with zope.browserresource could also be called out more clearly.

Multiple layers of that process can be overridden by other frameworks, starting with different/more specific registrations for ++resource. One framework that I work on doesn't use zope.browserresource at all, but does have objects that we use as resources.

(This is neither here nor there, but the reason I've been working on a lot of this documentation lately is that I've been implementing a [ZPT-based template system](http://ntinikola-chameleon.readthedocs.io/en/latest/) for the [nikola static website generator](https://getnikola.com/handbook.html). It's outside of any Zope3 framework, so I had to put a lot of these pieces together again for myself as I wrote and documented it; I'm *still* learning things as I help with this documentation---and your comments have been very helpful.)

mgedmin commented on src/zope/viewlet/viewlet.py:
```diff
@@ -149,12 +162,20 @@ class ResourceBundleViewletBase(object):
+
     def getResources(self):
+        """
+        Lookup all the resources in our desired paths using the
```
Not a native English speaker, but I think the verb form is "Look up".

jamadden replied:
Indeed, the dictionary agrees with you. That usage must be colloquial or dialectal. Thanks!

mgedmin replied:
FWIW there were other occurrences of this, but my headache said 'not now' and I didn't mark them out.

mgedmin replied:
Oh, I see you changed all three that I remember noticing.  Thank you!

jamadden replied:
I did a search and believe I corrected all the instances I introduced. There are still uses of 'lookup' as a verb in preexisting text in directives.rst, metadirectives.py and README.rst.

mgedmin commented on src/zope/viewlet/viewlet.py:
```diff
@@ -163,40 +184,45 @@ def render(self, *args, **kw):
+        """
+        Lookup all the resources for our desired items' paths using the
+        :class:`++resource++ namespace <zope.traversing.namespace.resource>`
+        and return a list of dictionaries.
```
It would be nice to document what those dictionaries contain.

mgedmin reviewed (approved):
I've quibbles, but this is definitely better than what we had before, and any minor imperfections can always be improved later.

jamadden committed:
Address review feedback.

Typo in readme.rst
Lookup -> retrieve
More about resources and calling them.
Document contents of a returned dictionary.

jamadden commented:
Thank you!
